Add GOES Data Download Manager Script

14Richa opened the pull request:
# Pull Request

## Description

This pull request adds a script for downloading GOES data. The script uses the `GOES2GO` library to download data for a specified time range and product. It includes functionality to save the downloaded data to a specified directory and log any errors encountered during the download process.

Fixes #222

## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [x] I have checked my code and corrected any misspellings

jacobbieker commented on satip/goes_download_manager.py:
```diff
@@ -0,0 +1,90 @@
+        while current_time <= end_time:
+            try:
+                # Download the data
+                ds = G.nearesttime(current_time)
```
Here, maybe checking if the data is already downloaded might be good?

jacobbieker commented on satip/goes_download_manager.py:
```diff
@@ -0,0 +1,90 @@
+                ds = G.nearesttime(current_time)
+
+                # Format the date string for filename
+                date_string = current_time.strftime("%Y-%m-%d_%H-%M-%S")
```
Rather than having the date string be from the `current_time`, it probably makes more sense to have it from the satellite image itself. The dataset should contain the acquisition time of the image, and that would be more accurate than this one. That would also make it easier to check if the data has already been downloaded.

jacobbieker commented on satip/goes_download_manager.py:
```diff
@@ -0,0 +1,90 @@
+
+        logging.info("Completed GOES data download.")
+
+if __name__ == "__main__":
```
This is great, and is a good test! If this part  could be split into its own test, and moved to the `tests/` folder, that would be great, so then its automatically run on all changes and we know if it breaks or not in the future.

jacobbieker commented on satip/goes_download_manager.py:
```diff
@@ -0,0 +1,90 @@
+            except Exception as e:
+                logging.error(f"Error downloading GOES data for {current_time}: {e}")
+
+            current_time += datetime.timedelta(minutes=1)
```
Maybe this can be changed depending on what product is selected? Or the domain? So that if we are going for the full disk, that is every 10 minutes, and we can add 10 minutes, rather than 1, and things like that? Just as a easy speed up to the process.

14Richa replied:
@jacobbieker  That's a valid point. Adapting the time increment based on the selected product or domain could indeed optimize the download process.

jacobbieker commented on satip/goes_download_manager.py:
```diff
@@ -0,0 +1,90 @@
+from goes2go import GOES
+
+
+class GOESDownloadManager:
```
This looks good! As mentioned in #222, I think the best would be for this to be integrated into the main DownloadManager, and a parameter that specifies the provider used to switch between EUMETSAT and Goes imagery.

jacobbieker reviewed (changes requested):
This is a good start! Thanks for this work. Its getting closer to what I was thinking for the image downloading. Mostly just some comments related to what was mentioned in #222, and some more simple checks to speed up the downloading and not download the same data multiple times.

14Richa commented:
> This is a good start! Thanks for this work. Its getting closer to what I was thinking for the image downloading. Mostly just some comments related to what was mentioned in #222, and some more simple checks to speed up the downloading and not download the same data multiple times.

Thanks for the comments @jacobbieker! I've addressed most of them and incorporated your suggestions. I've created a main DownloadManager class combining functionalities for EUMETSAT and GOES downloads. Also, I'm thinking of changing the filename from `eumetsat.py` to `download_manager.py` for better clarity. What do you think? Do you have any other suggestions for the filename?

jacobbieker commented on satip/eumetsat.py:
```diff
@@ -679,3 +687,63 @@ def eumetsat_cloud_name_to_datetime(filename: str):
+
+
+# Example usage
+if __name__ == "__main__":
```
This is good! For the example, I would put it in the docstring, or in the README instead. The code here in these files should just be the classes and functions. Just changing it like this will break other parts of the code, so either other ones should be updated to call the EUMETSATDownloadManager instead.

jacobbieker reviewed (changes requested):
Nice improvements! This kind of design is good, especially for being able to add more providers in the future. There will need to be some more changes to the code, as `DownloadManager` now doesn't do what it did before. So if you could update the other calls to be the `EUMETSATDownloadManager` that would be great! And might be the last changes for this PR I think.

jacobbieker commented:
> > This is a good start! Thanks for this work. Its getting closer to what I was thinking for the image downloading. Mostly just some comments related to what was mentioned in #222, and some more simple checks to speed up the downloading and not download the same data multiple times.
> 
> Thanks for the comments @jacobbieker! I've addressed most of them and incorporated your suggestions. I've created a main DownloadManager class combining functionalities for EUMETSAT and GOES downloads. Also, I'm thinking of changing the filename from `eumetsat.py` to `download_manager.py` for better clarity. What do you think? Do you have any other suggestions for the filename?

Yeah, making it `download_manager.py` is better, and moving the `EUMETSATDownloadManager` to a `eumetsat.py` file.

14Richa commented:
> Nice improvements! This kind of design is good, especially for being able to add more providers in the future. There will need to be some more changes to the code, as `DownloadManager` now doesn't do what it did before. So if you could update the other calls to be the `EUMETSATDownloadManager` that would be great! And might be the last changes for this PR I think.

Thanks for the feedback! I have updated the code as per your suggestion, changing the other calls to use `EUMETSATDownloadManager`.

jacobbieker commented on satip/eumetsat.py:
```diff
@@ -648,34 +677,3 @@ def create_and_download_datatailor_data(
-    return filesize_bytes / 1e6
-
-
-def eumetsat_filename_to_datetime(inner_tar_name):
```
Why were these ones removed?

14Richa replied:
Apologies, it was by mistake.

jacobbieker reviewed (commented):
Hi, one quick question left, then this I think is good to get merged! Did all the tests pass when you ran it?

jacobbieker commented:
Also, there will need to be a few updates to the tests, as they have changed in #229

14Richa commented:
> Hi, one quick question left, then this I think is good to get merged! Did all the tests pass when you ran it?

I've checked the tests, and `test_eumstat.py` is passing successfully. However, I encountered an error when running `test_utils.py`. I've attached the error message for your reference.

![image](https://github.com/openclimatefix/Satip/assets/41283476/614dd46c-6d33-465b-8883-4d229bf6f7fb)

I've also updated my branch according to the latest additions ( #229 )

14Richa commented:
Hey @jacobbieker, both of my tests are running successfully. Can we merge this PR now?

jacobbieker reviewed (approved):
Looks good! Thanks for this!

Rishikesh-Reddy commented on satip/goes_download_manager.py:
```diff
@@ -0,0 +1,96 @@
+            except Exception as e:
+                logging.error(f"Error creating directory {directory}: {e}")
+                raise
+    def download_goes_data(self, start_time, end_time, product='ABI-L1b-RadC',
```
It seems the product is simply named `ABI-L1b-Rad`, and the domain could be either `F`, `C`, or `M`. By default, the product should be labeled as `ABI-L1b-Rad`. It might be beneficial to revisit these details to confirm they align correctly.

peterdudfield commented:
@all-contributors please add @14Richa  for code
